data/tutorials/language/0lg_09_env.md

```diff
@@ -0,0 +1,611 @@
+```
+
+Here, f is a function that takes an argument x and returns another function. The returned function takes an argument y and returns yet another function. The final returned function takes an argument z and computes the result x + y + z.
+
+The consequence is that the type` int -> int -> int -> int` actually means `int -> (int -> (int -> int))`.
+
+Here is an example:
+
+```ocaml
+# ( + );;
+- : int -> int -> int = <fun>
+
+# let f : (int -> (int -> int)) = ( + );;
+val f : int -> int -> int = <fun>
+```
+
+In this example, the type of integer addition function `( + )` is displayed first. It prints as `int -> int -> int`, meaning it takes two integers and returns an integer. In the second statement, `f` is defined to have type `int -> (int -> int)`. Despite the type of `( + )` looking different, the binding with `f` is allowed because those two types are the same. The toplevel always prints the version without parentheses to make it easier to read.
+
+Putting the parentheses the other way does not work:
+```ocaml
+# let g : ((int -> int) -> int) = ( + );;
+Error: This expression has type int -> int -> int
+       but an expression was expected of type (int -> int) -> int
+       Type int is not compatible with type int -> int
+```
+
+In mathematical language, the type arrow operator _associates to the right_. Function types without parentheses should be treated as if they have parentheses placed to the right in the same way that the type of `f` was declared above. These types are the same:
+- `int -> int -> int`
+- `int -> (int -> int)`
+
+But only the first is displayed, so it is not the same as `(int -> int) -> int`.
+
+Most importantly, this means a *binary* function isn't really one that takes two parameters. In the case of addition, it is a function that takes an integer and returns a function of type `int -> int`. And it is indeed!
+```ocaml
+# ( + ) 2;;
+- : int -> int = <fun>
+```
+
+Passing a single integer to the addition returns a function of type `int -> int`. That value is a closure. The first argument, here 2, is captured as if it had been in an earlier definition.
+
+### Passing Tuples
+
+Consider this function:
+```ocaml
+# let space_cat (s, t) = s ^ " " ^ t;;
+val space_cat : string * string -> string = <fun>
+```
+
+It takes a pair of strings and returns a string.
+```ocaml
+# space_cat ("hello", "world");;
+- : string = "hello world"
+```
+
+Asking whether `space_cat` is a binary function or not makes sense. Morally, it takes two parameters: the two strings that sandwich the space character. Wrapped as a pair, they are a single piece of data, so `space_cat` has a single parameter. But it is that pair's elements that are relevant, not the pair as a whole. So morally, `space_cat` is a binary function. Syntactically, the call `space_cat ("hello", "world")` looks pretty much like a high school maths function call or a spreadsheet application, both having two parameters: `"hello"` and `"world"`.
```

Comment:
```suggestion
The function `space_cat` takes a single parameter, even though it looks like two arguments have been passed: "hello" and "world;" however, since they're enclosed in parentheses with a comma, they are actually a tuple. Tuples with multiple elements are a single value. 

In many imperative languages, this syntax reads as a function call with two parameters; but in OCaml, this syntax denotes applying the function `space_cat` to the tuple `("hello", "world")`. 
```